File: concentration.py
class Concentration:
    def __init__(self, n, t):
        self.N = n
        self.T = t
        self.acquiredCards = [0] * n
        self.P = 0

    def concentration(self, a, b, c, d):
        if self.T[a][b] == self.T[c][d]:
            self.acquiredCards[self.P] += 2
        else:
            self.P = -~self.P % self.N


H, W, N = map(int, input().split())
T = [input().split() for _ in range(H)]
concentration = Concentration(N, T)
for _ in range(int(input())):
    a, b, A, B = map(lambda x: ~-int(x), input().split())
    concentration.concentration(a, b, A, B)
for i in range(N):
    print(concentration.acquiredCards[i])

File: concentration.swift
import Foundation

class Concentration {
    private var N: Int
    private var T: [[String]]
    private var acquiredCards: [Int]
    private var P: Int

    init(n: Int, t: [[String]]) {
        self.N = n
        self.T = t
        self.acquiredCards = Array(repeating: 0, count: n)
        self.P = 0
    }

    func getAcquiredCard(_ i: Int) -> Int {
        return acquiredCards[i]
    }

    func concentration(a: Int, b: Int, c: Int, d: Int) {
        if T[a][b] == T[c][d] {
            acquiredCards[P] += 2
        } else {
            P = (P + 1) % N
        }
    }
}

func main() {
    let HWN = readLine()!.split(separator: " ").map { Int($0)! }
    let H = HWN[0], N = HWN[2]

    var T = [[String]]()
    for _ in 0..<H {
        T.append(readLine()!.split(separator: " ").map { String($0) })
    }

    let concentration = Concentration(n: N, t: T)

    let L = Int(readLine()!)!
    for _ in 0..<L {
        let input = readLine()!.split(separator: " ").map { Int($0)! - 1 }
        let (a, b, c, d) = (input[0], input[1], input[2], input[3])
        concentration.concentration(a: a, b: b, c: c, d: d)
    }

    for i in 0..<N {
        print(concentration.getAcquiredCard(i))
    }
}

main()
